File: frontend/src/lib/types/sentiment.ts
export interface FearGreedEntry {
  timestamp: string;
  value: number;
  classification: string;
}

export interface FearGreedResponse {
  current_value: number;
  current_class: string;
  signal: string;
  strength: number;
  trend: string;
  avg_7d: number;
  avg_30d: number;
  history: FearGreedEntry[];
}

export interface FundingRateEntry {
  exchange: string;
  symbol: string;
  rate: number;
  predicted_rate: number | null;
}

export interface FundingRateResponse {
  symbol: string;
  avg_rate: number;
  max_rate: number;
  min_rate: number;
  signal: string;
  strength: number;
  exchanges: FundingRateEntry[];
}

// ── Open Interest ─────────────────────────────────────────────

export interface OpenInterestEntry {
  timestamp: number;
  open_interest: number;
  open_interest_value: number;
}

export interface OpenInterestResponse {
  symbol: string;
  current_oi: number;
  current_oi_value: number;
  change_pct: number;
  signal: string;
  strength: number;
  trend: string;
  history: OpenInterestEntry[];
}

// ── Long/Short Ratio ─────────────────────────────────────────

export interface LongShortRatioEntry {
  timestamp: number;
  long_short_ratio: number;
  long_account: number;
  short_account: number;
}

export interface LongShortRatioResponse {
  symbol: string;
  current_ratio: number;
  long_pct: number;
  short_pct: number;
  avg_ratio: number;
  signal: string;
  strength: number;
  trend: string;
  history: LongShortRatioEntry[];
}

// ── Taker Buy/Sell Volume ────────────────────────────────────

export interface TakerVolumeEntry {
  timestamp: number;
  buy_sell_ratio: number;
  buy_vol: number;
  sell_vol: number;
}

export interface TakerVolumeResponse {
  symbol: string;
  current_ratio: number;
  buy_vol: number;
  sell_vol: number;
  avg_ratio: number;
  signal: string;
  strength: number;
  pressure: string;
  history: TakerVolumeEntry[];
}

File: backend/app/schemas/sentiment.py
from __future__ import annotations

from typing import Dict, List, Optional

from pydantic import BaseModel


class FearGreedEntry(BaseModel):
    timestamp: str
    value: int
    classification: str


class FearGreedResponse(BaseModel):
    current_value: int
    current_class: str
    signal: str
    strength: float
    trend: str
    avg_7d: float
    avg_30d: float
    history: List[FearGreedEntry]


class FundingRateEntry(BaseModel):
    exchange: str
    symbol: str
    rate: float
    predicted_rate: Optional[float] = None


class FundingRateResponse(BaseModel):
    symbol: str
    avg_rate: float
    max_rate: float
    min_rate: float
    signal: str
    strength: float
    exchanges: List[FundingRateEntry]


# ── Open Interest ─────────────────────────────────────────────


class OpenInterestEntry(BaseModel):
    timestamp: int
    open_interest: float
    open_interest_value: float


class OpenInterestResponse(BaseModel):
    symbol: str
    current_oi: float
    current_oi_value: float
    change_pct: float
    signal: str
    strength: float
    trend: str
    history: List[OpenInterestEntry]


# ── Long/Short Ratio ─────────────────────────────────────────


class LongShortRatioEntry(BaseModel):
    timestamp: int
    long_short_ratio: float
    long_account: float
    short_account: float


class LongShortRatioResponse(BaseModel):
    symbol: str
    current_ratio: float
    long_pct: float
    short_pct: float
    avg_ratio: float
    signal: str
    strength: float
    trend: str
    history: List[LongShortRatioEntry]


# ── Taker Buy/Sell Volume ────────────────────────────────────


class TakerVolumeEntry(BaseModel):
    timestamp: int
    buy_sell_ratio: float
    buy_vol: float
    sell_vol: float


class TakerVolumeResponse(BaseModel):
    symbol: str
    current_ratio: float
    buy_vol: float
    sell_vol: float
    avg_ratio: float
    signal: str
    strength: float
    pressure: str
    history: List[TakerVolumeEntry]
